build.sc

```diff
@@ -182,80 +122,138 @@ trait ChiselPublishModule extends CiReleaseModule {
+  override def scalacOptions = T {
+    if (!utils.isScala3(crossScalaVersion)) {
+      super.scalacOptions() ++ Agg("-Ymacro-annotations")
+    } else super.scalacOptions()
+  }
+}
 
-trait Firrtl extends common.FirrtlModule with CrossSbtModule with ScalafmtModule {
-  def millSourcePath = super.millSourcePath / os.up / "firrtl"
+trait HasScala2Plugin extends CrossSbtModule {
+  def pluginModule: Plugin
 
-  override def scalacOptions = v.commonOptions ++ Seq(
-    "-language:reflectiveCalls",
-    "-language:existentials",
-    "-language:implicitConversions",
-    "-Yrangepos", // required by SemanticDB compiler plugin
-    "-Xsource:3",
-    "-Xsource-features:infer-override"
-  )
+  override def scalacOptions = T {
+    if (!utils.isScala3(crossScalaVersion)) {
+      super.scalacOptions() ++ Agg(s"-Xplugin:${pluginModule.jar().path}")
+    } else super.scalacOptions()
+  }
 
-  def osLibModuleIvy = v.osLib
+  override def scalacPluginClasspath = T {
+    if (!utils.isScala3(crossScalaVersion)) {
+      super.scalacPluginClasspath() ++ Agg(pluginModule.jar())
+    } else super.scalacPluginClasspath()
+  }
+}
 
-  def json4sIvy = v.json4s
+object firrtl extends Cross[Firrtl](v.scalaCrossVersions)
 
-  def dataclassIvy = v.dataclass
+trait Firrtl extends CrossSbtModule with Cross.Module[String] with HasScala2MacroAnno with ScalafmtModule {
+  def millSourcePath = super.millSourcePath / os.up / "firrtl"
+  def scalaVersion = crossScalaVersion
+
+  override def scalacOptions = T {
+    if (utils.isScala3(crossScalaVersion)) {
+      Seq.empty[String]
+    } else {
+      v.scala2CommonOptions ++ Seq(
+        "-language:reflectiveCalls",
+        "-language:existentials",
+        "-language:implicitConversions",
+        "-Yrangepos", // required by SemanticDB compiler plugin
+        "-Xsource:3",
+        "-Xsource-features:infer-override"
+      )
+    }
+  }
 
-  def commonTextIvy = v.commonText
+  val commonDeps = Agg(
+    v.scopt,
+    v.commonText,
+    v.osLib,
+    v.json4s
+  )
 
-  def scoptIvy = v.scopt
+  def ivyDeps = if (utils.isScala3(crossScalaVersion)) {
+    commonDeps
+  } else {
+    commonDeps ++ Agg(v.dataclass)
+  }
 
   object test extends SbtModuleTests with TestModule.ScalaTest with ScalafmtModule {
     def ivyDeps = Agg(v.scalatest, v.scalacheck)
   }
 }
 
 object svsim extends Cross[Svsim](v.scalaCrossVersions)
-
-trait Svsim extends common.SvsimModule with CrossSbtModule with ScalafmtModule {
+trait Svsim extends CrossSbtModule with ScalafmtModule {
   def millSourcePath = super.millSourcePath / os.up / "svsim"
 
-  override def scalacOptions = v.commonOptions ++ Seq(
-    "-Xsource:3",
-    "-Xsource-features:case-apply-copy-access"
-  )
+  override def scalacOptions = T {
+    if (utils.isScala3(crossScalaVersion)) {
+      Seq.empty[String]
+    } else {
+      v.scala2CommonOptions ++ Seq(
+        "-Xsource:3",
+        "-Xsource-features:case-apply-copy-access"
```

Comment:
why add this feature?

Reply:
this is needed in Scala3 if you make the accessors private, here it avoids errors with Xsource:3